src/absil/illib.fs

```diff
@@ -672,42 +672,44 @@ let _ = eventually { use x = null in return 1 }
 //---------------------------------------------------------------------------
 // generate unique stamps
 //---------------------------------------------------------------------------
-
-type UniqueStampGenerator<'T when 'T : equality>() = 
-    let encodeTab = new Dictionary<'T,int>(HashIdentity.Structural)
-    let mutable nItems = 0
-    let encode str = 
-        if encodeTab.ContainsKey(str)
-        then
-            encodeTab.[str]
-        else
-            let idx = nItems
-            encodeTab.[str] <- idx
-            nItems <- nItems + 1
-            idx
-    member this.Encode(str)  = encode str
+[<Sealed>]
+type UniqueStampGenerator<'T when 'T : equality>(?initialCapacity) =
+    let encodeTab =
+        let initialCapacity = match initialCapacity with None -> 10 | Some c -> c
+        Dictionary<'T,int>(initialCapacity, HashIdentity.Structural)
+
+    member __.Encode(str) =
+        lock encodeTab <| fun () ->
+            let mutable strIdx = Unchecked.defaultof<_>
+            if encodeTab.TryGetValue (str, &strIdx) then strIdx
+            else
+                let idx = encodeTab.Count
+                encodeTab.Add (str, idx)
+                idx
 
 //---------------------------------------------------------------------------
 // memoize tables (all entries cached, never collected)
 //---------------------------------------------------------------------------
-    
-type MemoizationTable<'T,'U>(compute: 'T -> 'U, keyComparer: IEqualityComparer<'T>, ?canMemoize) = 
-    
-    let table = new System.Collections.Generic.Dictionary<'T,'U>(keyComparer) 
-    member t.Apply(x) = 
-        if (match canMemoize with None -> true | Some f -> f x) then 
-            let mutable res = Unchecked.defaultof<'U>
-            let ok = table.TryGetValue(x,&res)
-            if ok then res 
-            else
-                lock table (fun () -> 
-                    let mutable res = Unchecked.defaultof<'U> 
-                    let ok = table.TryGetValue(x,&res)
-                    if ok then res 
+[<Sealed>]
+type MemoizationTable<'T,'U>(compute: 'T -> 'U, keyComparer: IEqualityComparer<'T>, ?canMemoize, ?initialCapacity) =
```

Comment:
Likewise here - remove the canMemoize argument since it is not used (as far as I can see?)